packages/editor/src/dataviews/types.ts

```diff
@@ -0,0 +1,17 @@
+/**
+ * WordPress dependencies
+ */
+import type { AnyItem } from '@wordpress/dataviews';
+
+type PostStatus =
+	| 'published'
+	| 'draft'
+	| 'pending'
+	| 'private'
+	| 'future'
+	| 'auto-draft'
+	| 'trash';
+
+export interface Post extends AnyItem {
+	status?: PostStatus;
```

Comment:
@mcsf @jsnajdr  @sirreal This is not ideal. I was forced to make the status optional but in reality it's mandatory in "Post" objects. The reason I was forced to do that is that otherwise `Action<Post>` would give me an error saying `status` is not always present in `AnyItem`. I think this highlights something wrong in our "Action" types but I'm not really sure how to solve that.

Reply:
That said, from what I know of TS, I'm not sure `Action` is necessarily the culprit, but maybe `Post`.

```ts
interface Post extends AnyItem {
	status?: PostStatus;
}

type AnyItem = Record< string, any >;
```

Extending an interface A into an interface B means something straightforward: you declare a new interface which inherits the properties of its parent. But what does it mean to "extend" a custom record type (`AnyItem`) into an _interface_? The semantics of an interface are stricter than that of a record.

Reply:
> Extending an interface A into an interface B means something straightforward: you declare a new interface which inherits the properties of its parent. But what does it mean to "extend" a custom record type (`AnyItem`) into an _interface_? The semantics of an interface are stricter than that of a record.

After playing a bit in the TS Playground, I withdraw this suspicion. :)

Reply:
`B extends A` means that `B` is a subtype of `A`. That means that a value of type `B` can be safely used everywhere `A` is required.

Extending classes in classical OOP is one example of subtypes. If type `A` has a `a: string` field, and `B` extends that with `b: string` field, then `B` has both `a` and `b` fields. And `B` can be used in place of `A` because it has a compatible `a` field.

Another example of a subtype is when `'a' | 'b'` is a subtype of `string`. Of course, if a `string` is expected somewhere, surely you can use `'a'` or `'b'` there. But it seems paradoxical because `'a' | 'b'` is hardly an extension of `string`. Quite vice versa, the `'a' | 'b'` type is much smaller than `string`!

But then you need to realize that the set of all possible `B`s is also much smaller than the set of `A`s. Because `B` is a subset of `A`: it's a set of records that also have the `b` field in addition to `a`.

I think that the `extends` keyword is a bit confusing. It comes from classic OOP languages where classes and subclasses are the only usage of subtypes. But it feels counterintuitive when applied to other types.

Reply:
Thanks for taking the time to explain things. :) I did know these things, but it's always useful to check our assumptions, and I agree with you that "extends" was an interesting choice of word. I conjectured a bit around `Post` but my suspicion was premature.

Reply:
@youknowriad It gets fixed for me when `ActionButton` extends from `ActionBase<Item>` instead of `ActionBase<AnyItem>`. It was probably a typo because the `ActionModal` does the extension correctly.

Reply:
In the error message for `permanentlyDeletePostAction`, this part is the most important:
```
Types of property 'isEligible' are incompatible.
      Type '({ status, type }: Post) => boolean' is not assignable to type '(item: AnyItem) => boolean'.
```
The `isEligible` method is on the `ActionBase` type, and if the type is `ActionBase<AnyItem>`, the type of `isEligible` is:
```ts
( item: AnyItem ) => boolean
```
Because `Action<Post>` extends from `ActionBase<AnyItem>`, the `isEligible` method is expected to accept any `item` that has type `AnyItem`.

But you're assigning to `permanentlyDeletePostAction` an object whose `isEligible` is of a narrower type:
```ts
isEligible( { status, type }: Post )
```
This function doesn't accept an argument of type `AnyItem`. Because on `AnyItem` the `status` field can be anything, it can be `undefined` or anything else. But your `isEligible` expects it to be `PostStatus` and nothing else.

This gets fixed when `Action<Post>` extends from `ActionBase<Post>` because then the type of the `isEligible` `item` parameter is also `Post`.

I'm surprised that TypeScript was happy with the optional `status?: PostStatus`. Because that's still incompatible with `AnyItem` because `AnyItem` can have `status` value of type `number`, which is not compatible with `status?: PostStatus`.

Reply:
[This example](https://www.typescriptlang.org/play/?target=99&noImplicitOverride=true&ts=5.4.5#code/C4TwDgpgBAggdiAksCBbKBeKAlCBjAewCcATAHgGdgiBLOAcwBooBDBAPgG4AobulIgDMWeaAAUCVKBAAeKOCQqwEyNFADe3KFCotgAVwoAuHdTr0eAX178IQkdABCLChDIAVdhq1QaFAKIANjT0NABGgRAAFDQoqEbuAJRGYQQEkWxWvIRwUmEuECbOrmTwSHFeWJrafkEh4ZEmMXFGElSJmF6xaAB0ugYUfcAsRMAUAOqxABZRAOQQwUazidzW3PmuPbXBoRHR6v2GJnD6gYGWiZxAA) behaves in a way that I don't understand:
```ts
type AnyItem = Record<string, any>;

interface Post extends AnyItem {
  status: string;
}

interface Base<T> {
  isEligible(item: T): boolean;
}

const base: Base<AnyItem> = {
  isEligible: (item: Post) => item.status.startsWith('eli:')
}

base.isEligible({status: null});
```
In this program the types of `isEligible` don't match, it will crash at runtime (accessing `.startsWith` of `null`) and yet TypeScript doesn't report any type error. @sirreal Do you have any idea what's going on?

Reply:
Here's what I think is going on. It seem that a few things together create problems with the example.

- `AnyItem` is kind of a dangerous type. It says that it's an object where every string property has the value of type `any`. `any` is dangerous! If we don't just extend this type but actually use the type (like in `const base: Base<AnyItem>`), then `any` sneaks into the type system and wreaks havoc.
- `AnyItem` can be assigned to `Post`. This allows use to `Post` where we expect `AnyItem` (in the isEligible argument type).
- A type annotation `Post` is included on the argument `item` in `isEligible( item: Post )`. Because `AnyItem` is assignable to `Post`, this is allowed. If we'd used e.g. `string`, we'd see an error like this:
  > Types of parameters 'item' and 'item' are incompatible.
  >   Type 'AnyItem' is not assignable to type 'string'.

[I'd make a few changes:](https://www.typescriptlang.org/play/?target=99&noImplicitOverride=true&ts=5.4.5#code/C4TwDgpgBAggdiAksCBbKBeKAlCBjAewCcATAHgGdgiBLOAcwBooBXOAazgIHc4A+ANwAoIXRREAZgEM80RMjRQA3kKhQqU4CwoAuddTr1hagNoAPPVVoMAunracecIQF8RYiJJnQACgSpQEGYocCQUsAgK6CpqGlq6+tZGru5w4tKyUABCUhQQZAAqgcEQoeHwSCiofMqqUDQUAKIANjT0NABGzRAAFDRVOgUAlDodBATdUnDCbkKEcAEduRB6OXlkFVE1WDH1Ta3tXStQfVVDmDX9aAB0cdq3wFJEwBQA6v0AFj0A5BCtOt8hiklnlrg0Wm1Ot0eko7gk4Cxms0XENhHMCAtgFAQRA-FRVssyHjgNtampwQcocdTmhzhhLlUHpp7hpnm9Pj8-jQAUC3DjiWD9pCjjC4ToEUiUQIgA)

- `AnyItem` uses `Record<string, unknown>`. This prevents `any` types from sneaking in.
- The argument type annotation on `item` is removed from `base`. We'll now see a type error on `item.status` if using `Base<AnyItem>`.
- If we expect errors on `base.isEligible({ status: null })` then base should be typed `Base<Post>` instead of `Base<AnyItem>`.

```ts
type AnyItem = Record<string, unknown>;

interface IItem {
  status: string;
  [x: string]: unknown
}

interface Post extends AnyItem {
  status: string;
}

interface Base<T extends AnyItem> {
  isEligible(item:T):boolean;
}

const base: Base<AnyItem> = {
  isEligible: (item) => item.status.startsWith('eli:') // TYPE ERROR: 'item.status' if of type 'unknown'
}
base.isEligible({status: null});

const basePost: Base<Post> = {
  isEligible: (item) => item.status.startsWith('eli:')
}
basePost.isEligible({status:null}); // TYPE ERROR: 'null' is not assignable to type 'string'
```

Reply:
I was reflecting further and trying to produce a more succinct, direct, and clear example and discovered something fascinating.


I'm able to fix your example with one small change to the `Base` interface:

```diff
 interface Base<T> {
-  isEligible(item: T): boolean;
+  isEligible: (item: T) => boolean;
 }
```

This seems to work because the function type becomes an atomic type, so when we try to assign `(item: Post) => boolean` to `(item: AnyItem) => boolean`, it shows a type error. Before, it was checking compatibility of the argument type. I can't really explain this or why that apparently superficial change would significantly change the behavior of the type system, I think it may be a bug.

<details>

<summary>Some tricks I tried to get a minimal reproducible example</summary>

I tried using a function type alias and function interface. They are slightly different but both give a type error:

```ts
type AnyItem = Record<string, any>;

interface Post extends AnyItem { status: string; }

type F<T> = (item: T) => boolean;
interface IF<T> { (item: T): boolean; }

interface Base<T> {
  isEligible: F<T>
  isEligible2: IF<T>
}

const base: Base<AnyItem> = {
  isEligible: (item: Post) => item.status.startsWith('eli:'),
/*
Type '(item: Post) => boolean' is not assignable to type 'F<AnyItem>'.
  Types of parameters 'item' and 'item' are incompatible.
    Property 'status' is missing in type 'AnyItem' but required in type 'Post'.
*/ 
  isEligible2: (item: Post) => item.status.startsWith('eli:'),
/*
Type '(item: Post) => boolean' is not assignable to type 'IF<AnyItem>'.
  Types of parameters 'item' and 'item' are incompatible.
    Type 'AnyItem' is not assignable to type 'Post'.
*/
}
```

</details>

---

This statement from my previous comment is incorrect:

> `AnyItem` can be assigned to `Post`. This allows use to `Post where we expect `AnyItem` (in the `isEligible` argument type).

It's easy enough to see that's not true:

```ts
declare var a: AnyItem;
const b: Post = a; // type error
```

What confused me was that as soon as we change this section in your example we get a type error:

```ts
const base: Base<AnyItem> = {
  isEligible: (item: string) => item.status.startsWith('eli:')
  // Type '(item: string) => any' is not assignable to type '(item: AnyItem) => boolean'.
  //  Types of parameters 'item' and 'item' are incompatible.
  //    Type 'AnyItem' is not assignable to type 'string'
}
```

If the error is that `AnyItem` is not assignable to `string`, when there's no error with `Post` we'd assume that `AnyItem` is assignable to `Post`, but we've just seen that isn't the case…

Reply:
> I still don't understand the AnyItem / Item extends AnyItem abstraction

I think we could remove the `AnyItem` type completely. And whenever there is a `Foo< Item extends AnyItem >` type parameter, make it just `Foo< Item >`, without any constraints.

Because what do the generic types (`Action<Item>`, `ViewProps<Item>`) really do with the `Item` type? They put it into arrays and callback parameters. You can do that with any type. The types like `Action` can be completely generic.

Constraining the `Item` type to objects, and forbidding strings for example, is nice, but not very valuable IMO.

Reply:
I'm going to apply @jsnajdr's fix and we can consider removing AnyItem separately.

Reply:
I agree, I think that's a good idea.

Reply:
There's a lint rule to prevent using the weaker syntax [mentioned in this comment](https://github.com/WordPress/gutenberg/pull/62647#discussion_r1647980168), I'll proposed adding that rule in https://github.com/WordPress/gutenberg/pull/62718.

Reply:
I tried to remove `AnyItem` locally and it fails at three places:
1. `DataViews<Item>` has a `getItemId` prop which is a function of type `Item => string`. The prop is optional and defaults to `defaultGetItemId` which is like `item => item.id`. It works only when `Item` has a `id: string` field. Not for generic `Item`. To fix this, we must encode that `getItemId` is optional only when `Item` is a subtype of `{ id: string }`, otherwise you must supply a custom implementation.
2. Similar problem with defaults is in `normalizeFields`. There is a line:
   ```js
   const getValue = field.getValue || ( ( { item } ) => item[ field.id ] );
   ```
   Again, the default works only for specific subtypes where `Item` is a record that has a field that matches the `field.id` value. It's a type error to create a `Field<Item>` where `getValue` is fully optional.
3. In `ViewList<Item>` there is a line:
   ```js
   const selectedItem = data.findLast( ( item ) => selection.includes( item.id ) );
   ```
   Looking at `item.id` is not generic enough. The `getItemId` function should be used here.

Other than that, removing `AnyItem` works very well. And it turns out that the `Record<string,any>` construct was hiding some errors, because `any` tends to turn off typechecking. Now it's revealed that the `Field<Item>.id` should match the `Item`, and that it's not always safe to access `item.id`.

I pushed the branch, the commit id is a5cf5a4b1ca, you can have a look.

Maybe in the end we'll end up not removing `AnyItem`, but merely transforming it into `{ id: string }` base type, making `id` mandatory for all `Item`s.

Reply:
>  but merely transforming it into { id: string } base type, making id mandatory for all Items.

Is it possible somehow to say items should either have `id` string  or provide a getItemId prop.

Reply:
I tried out this and it works:
```tsx
type ItemWithId = { id: string };

type DataViewsProps< Item > = {
  /* ... */
} & ( Item extends ItemWithId
  ? { getItemId?: ( item: Item ) => string }
  : { getItemId: ( item: Item ) => string } );

const defaultGetItemId = ( item: ItemWithId ) => item.id;

Reply:
I managed to fix all three type errors I mentioned above, here's the draft PR: #62856.